doc/developer/design/20230216_role_based_access_control.md

```diff
@@ -0,0 +1,384 @@
+| `UPDATE`                             | `OBJECT(w)` usually `CLUSTER(U)`, `OBJECT(r)`                               |
+
+Superusers can do anything in the above table.
+
+In order to execute a read, the role only needs `r` permission on the direct objects being read. For example, a role can
+read from a view if it has `r` privileges on that view even if it does not have `r` privileges on the underlying objects
+within that view.
+
+We will add the following SQL commands:
+
+- `GRANT <privilege> ON <object> TO [ GROUP ] <role>`
+    - Gives `<privilege>` on `<object>` to `<role>`.
+    - Only the owner of `<object>` can grant privileges on it.
+        - Note: PostgreSQL allows other roles to grant privileges with the `WITH GRANT OPTION` option.
+    - `GROUP` is ignored.
+- `GRANT ALL [ PRIVILEGES ] ON <object> TO [ GROUP ] <role>`
+    - Same as grant above, but for all privileges.
+    - `PRIVILEGES` is ignored.
+- `REVOKE <privilege> ON <object> FROM [ GROUP ] <role>`
+    - Revokes `<privilege>` on `<object>` from `<role>`.
+    - Only the owner of `<object>` can revoke privileges from it.
+        - Note: PostgreSQL allows other roles to revoke privileges with the `WITH GRANT OPTION` option.
+    - `GROUP` is ignored.
+- `REVOKE ALL [ PRIVILEGES ] ON <object> FROM [ GROUP ] <role>`
+    - Same as revoke above but for all privileges.
+    - `PRIVILEGES` is ignored.
+- `ALTER <object_type> <object_name> OWNER TO <new_owner>`
+    - Transfers ownership of `<object_name>` to `<new_owner>`.
+    - Can only be run by the current owner (or member of owning role) or a superuser.
+    - Requires membership of `<new_owner>`.
+    - Requires `CREATE` privilege on the schema where `<object_name>` resides if the object resides in a schema.
+        - Rationale is that this is equivalent to `DROP` then `CREATE`.
+    - Requires `CREATE` privilege on the database where `<object_name>` resides if the object is a database.
+- `REASSIGN OWNED BY <old_role> TO <new_role>`
+    - Transfers ownership of all objects owned by `<old_role>` to `<new_role>`.
+    - Can only be run by a member of `<old_role>` and `<new_role>` or a superuser.
+    - Requires `CREATE` privilege on all schemas and databases where all objects reside.
+        - TODO: This isn't explicitly stated in th docs, but would make sense base on the `ALTER` privileges. I will
+          double-check this.
+    - In PostgreSQL, this only affects the current database. We will diverge here and have it affect all databases.
+- `DROP OWNED BY <name> [ CASCADE | RESTRICT]`
+    - Drops all objects owned by `<name>`.
+    - Requires membership of `<name>`.
+    - In PostgreSQL, this only affects the current database. We will diverge here and have it affect all databases.
+    - Revokes all privileges granted to `<name>`.
+    - Does not drop owned databases or clusters.
+    - Default is `RESTRICT`.
```

Comment:
I might skip these two to start, because they are 1) annoying to implement, 2) not that often used, and 3) it's possible (though annoying) to work around their absence.

Reply:
I just moved these into a 5th optional phase.